### DISCUSSION TITLE
Raman Tensor Analysis

### DISCUSSION THREAD

**Discussion by MouadBik** [nguyen-group/QERaman#10] [Q&A] [answered]
Dear developers,

I want to express my gratitude for the excellent code you have provided. I have successfully executed the examples and achieved results consistent with those presented in the paper. Currently, I am endeavoring to analyze silicon bulk material to replicate experimental findings (first, does the code also support 3D materials?). However, I have encountered some queries that necessitate your assistance:

1. Does the parameter `nbnd` in `scf.in` influence the final Raman tensor? Should we conduct a convergence analysis regarding its value? Alternatively, is it advisable to maintain `nbnd` equal to the number of electrons in the system (as indicated in `scf.out`)?
2. In the output file `raman.out`, multiple Raman tensors are listed for various "iel" values. Which values should be prioritized for consideration?
3. Upon completing the convergence of k-points based on the a and b values from the Raman tensor extracted in `raman.out`, what methodology should be employed to plot Raman spectra for arbitrary configurations?

Additionally, I would like to propose a suggestion. I have successfully compiled the code using the GPU-enabled version of Quantum Espresso. However, the results appear anomalous. Could you confirm whether the code supports GPU acceleration or if there are forthcoming plans to incorporate such functionality?

Thank you for your attention to these matters.

**Comment by nguyen-group** [comment, marked as answer]
Hi,

Thank you for being so interested in our code.

>> Does the parameter nbnd in scf.in influence the final Raman tensor? Should we conduct a convergence analysis regarding its value? Alternatively, is it advisable to maintain nbnd equal to the number of electrons in the system (as indicated in scf.out)?

nbnd will affect the final Raman tensor. A convergence checking should be done for nbnd parameter.

>> In the output file raman.out, multiple Raman tensors are listed for various "iel" values. Which values should be prioritized for consideration?

iel is the laser energy, which is an input parameter in the experiment. So it depends on the experiment data that you want to compare to select the laser energy value.

>> Upon completing the convergence of k-points based on the a and b values from the Raman tensor extracted in raman.out, what methodology should be employed to plot Raman spectra for arbitrary configurations?

As shown in our published paper: [doi.org/10.1016/j.cpc.2023.108967](https://doi.org/10.1016/j.cpc.2023.108967) you can use equation 7 to plot the Raman intensity from the Raman tensor for arbitrary configurations.

>> Additionally, I would like to propose a suggestion. I have successfully compiled the code using the GPU-enabled version of Quantum Espresso. However, the results appear anomalous. Could you confirm whether the code supports GPU acceleration or if there are forthcoming plans to incorporate such functionality?

Our main code is based on the electron-photon and electron-phonon matrix elements from the QE. We did not test with the GPU version of QE. I think if the QE supports the electron-photon and electron-phonon in GPU, the code should be okay. We would appreciate it if you could compare the electron-photon and electron-phonon matrix elements from the GPU and non-GPU versions. In the case that they are the same, I believe that the Raman tensor should be the same. If they are not the same, it will not be easy for us to develop the GPU version. 

Best,
Nguyen

**Comment by MouadBik** [comment]
Dear Nguyen,

Firstly, I extend my gratitude for your prompt and insightful response to my inquiries. Your guidance has been instrumental in advancing my understanding of the intricacies involved.

In an attempt to explore the feasibility of GPU acceleration with QERaman, I conducted a small-scale test using bulk cubic silicon. Enclosed within this correspondence are the frequencies extracted from both the CPU and GPU versions for your perusal:

### Results:
#### CPU Version:
**************************************************************************
     freq (    1) =      -0.019412 [THz] =      -0.647498 [cm-1]
     freq (    2) =      -0.018256 [THz] =      -0.608940 [cm-1]
     freq (    3) =      -0.017489 [THz] =      -0.583356 [cm-1]
     freq (    4) =      15.675674 [THz] =     522.884194 [cm-1]
     freq (    5) =      15.675675 [THz] =     522.884218 [cm-1]
     freq (    6) =      15.675676 [THz] =     522.884278 [cm-1]
**************************************************************************

#### GPU Version:
**************************************************************************
     freq (    1) =     -56.979275 [THz] =   -1900.624034 [cm-1]
     freq (    2) =     -56.979273 [THz] =   -1900.623966 [cm-1]
     freq (    3) =     -56.979273 [THz] =   -1900.623948 [cm-1]
     freq (    4) =     -52.442828 [THz] =   -1749.304436 [cm-1]
     freq (    5) =     -52.442827 [THz] =   -1749.304409 [cm-1]
     freq (    6) =     -52.442822 [THz] =   -1749.304234 [cm-1]
**************************************************************************

As evident from the provided data, there exists a notable disparity between the frequencies obtained from the CPU and GPU versions. Consequently, it appears that achieving consistency between these two implementations may present a significant challenge.

Thank you once again for your invaluable assistance and guidance.

Best regards,

Mouad Bikerouin

**Reply by nguyen-group** [reply]
Dear Mouad Bikerouin

Thank you for performing calculations on GPU. It seems that the Phonon does not work well in GPU. 
We might need to wait for a stable version of GPU from the QE developer. If the band.x and ph.x do not work well, we can do nothing with QERaman.

Best regards,
Nguyen
